tests/test_pyth_client.py

```diff
@@ -0,0 +1,378 @@
+    return PythPriceAccount(
+        key=SolanaPublicKey(BCH_PRICE_ACCOUNT_KEY_DEVNET),
+        solana=solana_client,
+    )
+
+
+@ pytest.fixture
+def price_account_bytes() -> bytes:
+    return base64.b64decode(PRICE_ACCOUNT_B64_DATA)[_ACCOUNT_HEADER_BYTES:]
+
+
+@pytest.fixture()
+def mock_get_account_info(mocker: MockerFixture) -> AsyncMock:
+    async_mock = AsyncMock(side_effect=get_account_info_resp)
+    mocker.patch('pythclient.solana.SolanaClient.get_account_info', side_effect=async_mock)
+    return async_mock
+
+
+@ pytest.fixture()
+def mock_get_program_accounts(mocker: MockerFixture) -> AsyncMock:
+    async_mock = AsyncMock(side_effect=get_program_accounts_resp)
+    mocker.patch('pythclient.solana.SolanaClient.get_program_accounts', side_effect=async_mock)
+    return async_mock
+
+
+def test_products_property_not_loaded(pyth_client: PythClient):
+    with pytest.raises(NotLoadedException):
+        pyth_client.products
+
+
+@pytest.mark.asyncio
+async def test_get_products(
+    pyth_client: PythClient,
+    mock_get_account_info: AsyncMock,
+    product_account: PythProductAccount
+) -> None:
+    products = await pyth_client.get_products()
+    for i in range(len(products)):
+        assert products[i].key == product_account.key
+
+
+def test_get_ratelimit(
+    pyth_client: PythClient,
+) -> None:
+    ratelimit = pyth_client.solana_ratelimit
+    assert ratelimit._get_overall_interval() == 0
+    assert ratelimit._get_method_interval() == 0
+    assert ratelimit._get_connection_interval() == 0
+
+
+@pytest.mark.asyncio
+async def test_get_mapping_accounts(
+    pyth_client: PythClient,
+    mock_get_account_info: AsyncMock,
+    mapping_account: PythMappingAccount
+) -> None:
+    mapping_accounts = await pyth_client.get_mapping_accounts()
+    assert len(mapping_accounts) == 1
+    assert mapping_accounts[0].key == mapping_account.key
+
+    # call get_mapping_accounts again to test pyth_client returns self._mapping_accounts since
+    # it has been populated due to previous call
+    mapping_accounts = await pyth_client.get_mapping_accounts()
+    assert len(mapping_accounts) == 1
+    assert mapping_accounts[0].key == mapping_account.key
+
+
+@pytest.mark.asyncio
+async def test_get_all_accounts(
+    pyth_client: PythClient,
+    mock_get_account_info: AsyncMock,
+    mapping_account: PythMappingAccount,
+    mapping_account_entries: List[SolanaPublicKey],
+    product_account: PythProductAccount,
+    price_account: PythPriceAccount,
+    product_account_bytes: bytes,
+    price_account_bytes: bytes
+) -> None:
+    products = await pyth_client.get_products()
+    for product in products:
+        product.update_from(buffer=product_account_bytes, version=_VERSION_2)
+        prices = await product.get_prices()
+        for price in prices.values():
+            price.update_from(buffer=price_account_bytes, version=_VERSION_2)
+
+    accounts = await pyth_client.get_all_accounts()
```

Comment:
You know the order that these will be returned since you mock things, right? If so, why not just make assertions against accounts[0], accounts[1], accounts[2], etc?

That way you don't need the `isinstance()` blocks below. Where possible, always err on the side of making tests less code and more static data. Be explicit for correctness. Your isinstance conditional doesn't have an else block at all, so if there was some other type of erroneous account, this code would happily pass even if it was bad data.

Reply:
good catch, thank you for the suggestion, makes sense to use static data when testing